### PR TITLE
Auth logic relation declaration AST datastructure

### PR DESCRIPTION
**Pull request by harsha-mandadi**
This PR constructs a Relation Declaration AST node. Lowering ast for relation declaration will be handled in another PR keep the PR size small.

### CODE REVIEW

File: src/ir/auth_logic/ast.h [hunk 1]
```diff
@@ -152,6 +152,43 @@ class Assertion {
...
+  const std::string& predicate_name() const { return predicate_name_; }
+  const bool& is_attribute() const { return is_attribute_; }
+  const std::vector<RelationDeclarationType>& arg_type() const {
+    return arg_type_;
+  }
+
+ private:
+  std::string predicate_name_;
+  bool is_attribute_;
+  std::vector<RelationDeclarationType> arg_type_;
```
Comment (markww) [inline src/ir/auth_logic/ast.h:189]: It is odd to call a field containing multiple argument types in a vector `arg_type_` as if it were singular. Perhaps this should be rewritten to `argument_types_`.

File: src/ir/auth_logic/ast.h [hunk 2]
```diff
@@ -186,16 +223,22 @@ class Query {
...
  public:
-  explicit Program(std::vector<SaysAssertion> says_assertions,
+  explicit Program(std::vector<RelationDeclaration> relation_declaration,
+                   std::vector<SaysAssertion> says_assertions,
                    std::vector<Query> queries)
-      : says_assertions_(std::move(says_assertions)),
+      : relation_declaration_(std::move(relation_declaration)),
+        says_assertions_(std::move(says_assertions)),
         queries_(std::move(queries)) {}
+  const std::vector<RelationDeclaration>& relatin_declaration() const {
```
Comment (markww) [inline src/ir/auth_logic/ast.h:232]: typo: should be `relation_declaration`.

File: src/ir/auth_logic/ast_construction.cc [hunk 3]
```diff
@@ -33,8 +33,8 @@
 #include "src/policy/auth_logic/auth_logic_cc_generator_grammar.inc/auth_logic_cc_generator/AuthLogicLexer.h"
 #include "src/policy/auth_logic/auth_logic_cc_generator_grammar.inc/auth_logic_cc_generator/AuthLogicParser.h"
 
-using auth_logic_cc_generator::AuthLogicParser;
 using auth_logic_cc_generator::AuthLogicLexer;
+using auth_logic_cc_generator::AuthLogicParser;
```
Comment (markww) [inline src/ir/auth_logic/ast_construction.cc:37]: These `using` declarations should be inside the `namespace`.

File: src/ir/auth_logic/ast.h [hunk 4]
```diff
@@ -152,6 +152,43 @@ class Assertion {
...
+// 1. NumberType
+// 2. PrincipalType
+// 3. User defined CustomType (string to store name of the type)
+class RelationDeclarationType {
+ public:
+  using RelationDeclarationTypeVariantType =
+      std::variant<constantType, std::string>;
+  explicit RelationDeclarationType(RelationDeclarationTypeVariantType value)
+      : value_(std::move(value)) {}
+  const RelationDeclarationTypeVariantType& GetValue() const { return value_; }
```
[the remark this answers is not in the document]
Comment (bgogul) [reply src/ir/auth_logic/ast.h:167]: In terms of naming, the following might also work: `AuthLogicType` or `VariableType` or `ArgumentType`. We also use this for Attribute, so `RelationArgumentType` or `RelationParameterType` might not be quite accurate. 

Anyhow,. assuming `ArgumentType` as the name, you can simplify this datastructure without variants:

```C++
class ArgumentType {
  enum class Kind { kNumber, kPrincipal, kCustom};
private:
  Kind kind_;
  std::string name_;
};
```

**Review by bgogul — COMMENTED** [review]
It looks like you don't need `variant` at all.

File: src/ir/auth_logic/ast.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (harsha-mandadi) [reply src/ir/auth_logic/ast.h:167]: Ah, interesting. Nested class is a good idea. Thanks Gogul. I was looking for something simpler and found my answer.

File: src/ir/auth_logic/ast.h (hunk 2 above)
Comment (harsha-mandadi) [reply src/ir/auth_logic/ast.h:232]: Oops!

File: src/ir/auth_logic/ast_construction.cc [hunk 5]
```diff
@@ -97,12 +97,14 @@ static Query ConstructQuery(AuthLogicParser::QueryContext& context) {
...
 static Program ConstructProgram(AuthLogicParser::ProgramContext& context) {
+  std::vector<RelationDeclaration> relation_declarations;
   std::vector<SaysAssertion> says_assertions;
   std::vector<Query> queries = utils::MapIter<Query>(
       context.query(), [](AuthLogicParser::QueryContext* query_context) {
         return ConstructQuery(*CHECK_NOTNULL(query_context));
       });
-  return Program(std::move(says_assertions), std::move(queries));
+  return Program(std::move(relation_declarations), std::move(says_assertions),
+                 std::move(queries));
```
[the remark this answers is not in the document]
Comment (harsha-mandadi) [reply src/ir/auth_logic/ast_construction.cc:107]: This is a temporary state and hence empty says_assertion and relation_declaration. This is a WIP PR and will add details for says_assertions and relation_declaration eventually.

File: src/ir/auth_logic/ast.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (harsha-mandadi) [reply src/ir/auth_logic/ast.h:167]: Question about std::visit for variant types. Our current ast.h has multiple variant types returning through getters. We should probably change them as well?

File: src/ir/auth_logic/ast.h [hunk 6]
```diff
@@ -152,6 +152,43 @@ class Assertion {
   AssertionVariantType value_;
 };
 
+enum class constantType { NumberType, PrincipalType };
```
[the remark this answers is not in the document]
Comment (harsha-mandadi) [reply src/ir/auth_logic/ast.h:155]: Noted!

File: src/ir/auth_logic/ast.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (markww) [reply src/ir/auth_logic/ast.h:167]: Yes, definitely. It's bad encapsulation to return these out of the class that contains them.

File: src/ir/auth_logic/ast.h [hunk 7]
```diff
@@ -139,6 +139,46 @@ class Assertion {
   AssertionVariantType value_;
 };
 
+enum class ConstantType { NumberType, PrincipalType };
```
Comment (bgogul) [inline src/ir/auth_logic/ast.h:142]: `kNumberType, kPrincipalType`

See https://google.github.io/styleguide/cppguide.html#Enumerator_Names

File: src/ir/auth_logic/ast.h [hunk 8]
```diff
@@ -139,6 +139,46 @@ class Assertion {
...
+
+// Relation declaration have types are of 3 forms
+// 1. NumberType
+// 2. PrincipalType
+// 3. User defined CustomType (string to store name of the type)
+class ArguementType {
```
Comment (bgogul) [inline src/ir/auth_logic/ast.h:148]: typo?  `ArguementType` -> `ArgumentType`

File: src/ir/auth_logic/ast.h [hunk 9]
```diff
@@ -139,6 +139,46 @@ class Assertion {
...
+ public:
+  enum class Kind { kNumber, kPrincipal, kCustom };
+  explicit ArguementType(Kind kind, std::string name)
+      : kind_(std::move(kind)), name_(std::move(name)) {}
+  explicit ArguementType(Kind kind) : kind_(std::move(kind)) {}
```
Comment (bgogul) [inline src/ir/auth_logic/ast.h:153]: Why is this needed? Don't we always want to have a name? It looks like you expect name only if kind is `CustomType`? An alternative way to do this with static factory methods and simply hide the constructor:

```C++
class ArgumentType {
public:
  // ...
  static ArgumentType MakePrincipalType() { return ArgumentType(Kind::kPrincipal, ""); }
  static ArgumentType MakeNumberType() { return ArgumentType(Kind::kNumber, ""); }
  static ArgumentType MakeCustomType(absl::string_view name) { return ArgumentType(Kind::kCustom, name); }

  ...
private:
     explicit ArgumentType(Kind kind, absl::string_view name) : kind_(std::move(kind)), name_(name) {}
}
```


```

File: src/ir/auth_logic/ast.h [hunk 10]
```diff
@@ -139,6 +139,46 @@ class Assertion {
...
+  const std::string& relation_name() const { return relation_name_; }
+  const bool& is_attribute() const { return is_attribute_; }
+  const std::vector<std::pair<std::string, ArguementType>>& arguements() const {
+    return arguements_;
+  }
+
+ private:
+  std::string relation_name_;
+  bool is_attribute_;
+  std::vector<std::pair<std::string, ArguementType>> arguements_;
```
Comment (bgogul) [inline src/ir/auth_logic/ast.h:179]: typo? `arguements_` -> `arguments_`.

File: src/ir/auth_logic/ast_construction.cc [hunk 11]
```diff
@@ -128,6 +128,40 @@ static Fact ConstructFact(std::forward_list<Principal> delegation_chain,
                        *CHECK_NOTNULL(can_say_fact_context.fact()));
 }
 
+static ArguementType ConstructArguementType(
```
Comment (bgogul) [inline src/ir/auth_logic/ast_construction.cc:131]: typo: `ArugmentType ConstructArugmentType(...`

File: src/ir/auth_logic/ast_construction.cc [hunk 12]
```diff
@@ -182,6 +216,14 @@ static SaysAssertion ConstructSaysAssertion(
 
 static Program ConstructProgram(
     AuthLogicParser::ProgramContext& program_context) {
+  std::vector<RelationDeclaration> relation_declarations =
+      utils::MapIter<RelationDeclaration>(
```
Comment (bgogul) [inline src/ir/auth_logic/ast_construction.cc:220]: nit: `auto relation_declarations =  utils::MapIter<RelationDeclaration>(...`

**Review by bgogul — COMMENTED** [review]
Left style-related comments...

File: src/ir/auth_logic/ast.h (hunk 9 above)
Comment (markww) [reply src/ir/auth_logic/ast.h]: +1. This has the additional benefit of being more type-safe, in the sense that you are disallowing the bad pattern of making a `CustomType` without a name or a `NumberType` with a name.

File: src/ir/auth_logic/ast.h [hunk 13]
```diff
@@ -139,6 +139,46 @@ class Assertion {
...
+ private:
+  Kind kind_;
+  std::string name_;
+};
+
+class RelationDeclaration {
+ public:
+  explicit RelationDeclaration(
+      std::string relation_name, bool is_attribute,
+      std::vector<std::pair<std::string, ArgumentType>> arguments)
```
Comment (markww) [inline src/ir/auth_logic/ast.h:166]: This use of `std::pair` is interesting; should there be a simple `Argument` class that contains within it the fields `std::string argument_name_` and `ArgumentType argument_type_`? It's usually the case that `std::pair` and `std::tuple` should only be used for locals or weird template-y situations. Usually for fields of classes and other more formal layout situations a more formal class should be used.

File: src/ir/auth_logic/ast_construction.cc [hunk 14]
```diff
@@ -128,6 +128,40 @@ static Fact ConstructFact(std::forward_list<Principal> delegation_chain,
...
+static RelationDeclaration ConstructRelationDeclaration(
+    AuthLogicParser::RelationDeclarationContext& relation_declaration_context) {
+  std::vector<std::pair<std::string, ArgumentType>> arguments;
+  for (uint64_t i = 0; i < relation_declaration_context.authLogicType().size();
+       ++i) {
+    arguments.push_back(std::pair(
+        relation_declaration_context.ID(i + 1)->getText(),
+        ConstructArgumentType(
+            *CHECK_NOTNULL(relation_declaration_context.authLogicType(i)))));
+  }
```
Comment (markww) [inline src/ir/auth_logic/ast_construction.cc:158]: This is pretty subtle and looks like it could use a comment explaining it. Why are we looking up `ID(i + 1)`? Is this a 1-indexed collection? It sets off my "maybe there's a bug" alarm bells to see `i+1` and `i` used to index into stuff in `relation_declaration_context` in the same loop, but it could be fine depending.

File: src/ir/auth_logic/ast_construction.cc [hunk 15]
```diff
@@ -128,6 +128,40 @@ static Fact ConstructFact(std::forward_list<Principal> delegation_chain,
...
+  auto& custom_type_context =
+      *CHECK_NOTNULL(dynamic_cast<AuthLogicParser::CustomTypeContext*>(
+          &auth_logic_type_context));
+  return ArgumentType(ArgumentType::Kind::kCustom,
+                      custom_type_context.ID()->getText());
+}
+
+static RelationDeclaration ConstructRelationDeclaration(
+    AuthLogicParser::RelationDeclarationContext& relation_declaration_context) {
+  std::vector<std::pair<std::string, ArgumentType>> arguments;
```
Comment (markww) [inline src/ir/auth_logic/ast_construction.cc:151]: If you know the size of a `vector` in advance (as you do here), it's a cheap and safe optimization to call `reserve` with the size that you expect the vector to be.

File: src/ir/auth_logic/ast.h (hunk 9 above)
Comment (harsha-mandadi) [reply src/ir/auth_logic/ast.h]: I made changes to ArgumentType Node to make it look simpler. The reason I left out the name is that, for NumberType and PrincipalType, it is understood that they correspond to 'Number' and 'Principal' (Handle this in ast_construction). Whereas for customType, we need to explicitly provide the name for type. As this does not look like a good pattern to have, I changed to having name for all types. PTAL.

File: src/ir/auth_logic/ast.h (hunk 13 above)
Comment (harsha-mandadi) [reply src/ir/auth_logic/ast.h]: Added a new node for Argument to remove std::pair

File: src/ir/auth_logic/ast_construction.cc (hunk 14 above)
Comment (harsha-mandadi) [reply src/ir/auth_logic/ast_construction.cc]: Comments provided in the file explaining the need to `i` and `i+1`.

File: src/ir/auth_logic/ast.h [hunk 16]
```diff
@@ -139,6 +139,55 @@ class Assertion {
...
+enum class ConstantType { kNumberType, kPrincipalType };
+
+// Relation declaration types are of 3 forms
+// 1. NumberType
+// 2. PrincipalType
+// 3. User defined CustomType (string to store name of the type)
+class ArgumentType {
+ public:
+  enum class Kind { kNumber, kPrincipal, kCustom };
+  explicit ArgumentType(Kind kind, std::string name)
```
Comment (bgogul) [inline src/ir/auth_logic/ast.h:151]: No need for `std::move` for kind. Kind is already cheap to copy.

optional: Instead of taking a string, you could take an absl::string_view:
```C++
explicit ArgumentType(Kind kind, absl::string_view name): kind_(kind), name_(name) {}
...
absl::string_view name() const { return name_; }

```
Note that  the field `  std::string name_; ` would still stay as string.

File: src/ir/auth_logic/ast.h [hunk 17]
```diff
@@ -139,6 +139,55 @@ class Assertion {
...
+      : kind_(std::move(kind)), name_(std::move(name)) {}
+  const Kind& kind() const { return kind_; }
```
Comment (bgogul) [inline src/ir/auth_logic/ast.h:153]: Simply `Kind` is sufficient here.

File: src/ir/auth_logic/ast.h [hunk 18]
```diff
@@ -139,6 +139,55 @@ class Assertion {
...
+  const std::string name() const { return name_; }
```
Comment (bgogul) [inline src/ir/auth_logic/ast.h:154]: `const std::string&` here. Otherwise, you will make an unnecessary copy. You can also use `absl::string_view` above.

File: src/ir/auth_logic/ast.h [hunk 19]
```diff
@@ -139,6 +139,55 @@ class Assertion {
...
+
+ private:
+  Kind kind_;
+  std::string name_;
+};
+
+class Argument {
+ public:
+  explicit Argument(std::string argument_name, ArgumentType argument_type)
```
Comment (bgogul) [inline src/ir/auth_logic/ast.h:163]: optional: `std::string argument_name` could be `absl::string_view` as in  `ArgumentType` case.

File: src/ir/auth_logic/ast.h [hunk 20]
```diff
@@ -139,6 +139,55 @@ class Assertion {
...
+      : argument_name_(std::move(argument_name)),
+        argument_type_(std::move(argument_type)) {}
+  const std::string& argument_name() { return argument_name_; }
```
Comment (bgogul) [inline src/ir/auth_logic/ast.h:166]: Add const declarations at the end, which marks the members as const functions. Also, return `ArgumentType` instead of `const ArgumentType`
```
  const std::string& argument_name() const  { return argument_name_; } 
  ArgumentType argument_type() const { return argument_type_; }
```

File: src/ir/auth_logic/ast.h [hunk 21]
```diff
@@ -139,6 +139,55 @@ class Assertion {
...
+
+class RelationDeclaration {
+ public:
+  explicit RelationDeclaration(std::string relation_name, bool is_attribute,
+                               std::vector<Argument> arguments)
+      : relation_name_(std::move(relation_name)),
+        is_attribute_(std::move(is_attribute)),
+        arguments_(std::move(arguments)) {}
+  const std::string& relation_name() const { return relation_name_; }
+  const bool& is_attribute() const { return is_attribute_; }
```
Comment (bgogul) [inline src/ir/auth_logic/ast.h:182]: simply `bool` instead of `const bool&`.

File: src/ir/auth_logic/ast.h [hunk 22]
```diff
@@ -139,6 +139,55 @@ class Assertion {
...
+  const ArgumentType argument_type() { return argument_type_; }
+
+ private:
+  std::string argument_name_;
+  ArgumentType argument_type_;
+};
+
+class RelationDeclaration {
+ public:
+  explicit RelationDeclaration(std::string relation_name, bool is_attribute,
```
Comment (bgogul) [inline src/ir/auth_logic/ast.h:176]: optional: absl::string_view.

File: src/ir/auth_logic/ast_construction.cc [hunk 23]
```diff
@@ -128,6 +128,50 @@ static Fact ConstructFact(std::forward_list<Principal> delegation_chain,
...
+                      custom_type_context.ID()->getText());
+}
+
+// grammar rule for relationDeclaration:
+// '.decl' ATTRIBUTE? ID '(' ID ':' authLogicType (',' ID ':'authLogicType)* ')'
+// When translated to Antlr generated Parse tree, ID[0] corresponds to
+// relation_name.
+// (ID[1] : authLogicType[0]), (ID[2] : authLogicType[1]),.....,(ID[n] :
+// authLogicType[n-1]) corresponds to
+//  arguements for the relation.
```
Comment (bgogul) [inline src/ir/auth_logic/ast_construction.cc:157]: nit: Move this into the body of the function?

**Review by bgogul — COMMENTED** [review]
Noticed a few C++ no nos. Can you take care of those please?

Otherwise, looks good.

**Review by bgogul — APPROVED** [review]
Nice. Looks good!
